### PR TITLE
Mv spin locks

### PR DESCRIPTION
**Pull request by matthewvon**
Details of the changes are here:  https://github.com/basho/leveldb/wiki/Mv-spin-locks

### CODE REVIEW

**Comment by bookshelfdave** [comment]
@matthewvon looks like this needs a rebase or some cleanup (from "We can’t automatically merge this pull request."). I can do some review tomorrow.

**Comment by matthewvon** [comment]
The issue is that I manually added .travis.yml to both this branch and another branch already on master.  The two files differed by a line feed at the end ... git is so wise.

File: db/db_impl.cc [hunk 1]
```diff
@@ -1564,7 +1564,7 @@ Status DBImpl::Write(const WriteOptions& options, WriteBatch* my_batch) {
       // throttle is per key write, how many in batch?
       count=(NULL!=my_batch ? WriteBatchInternal::Count(my_batch) : 1);
       env_->SleepForMicroseconds(throttle * count);
-      gPerfCounters->Add(ePerfDebug0, throttle);
+      gPerfCounters->Add(ePerfDebug0, throttle * count);
```
Comment (bookshelfdave) [inline db/db_impl.cc:1567]: is this related to spin locks, or just a perf counter bug fix?

File: port/port_posix.h [hunk 2]
```diff
@@ -102,6 +102,30 @@ class Mutex {
...
+
+#if defined(_POSIX_SPIN_LOCKS) && 0<_POSIX_SPIN_LOCKS
+class Spin {
+ public:
+  Spin();
+  ~Spin();
+
+  void Lock();
+  void Unlock();
+  void AssertHeld() { }
```
Comment (bookshelfdave) [inline port/port_posix.h:114]: I see AssertHeld() { } in other classes, what's the point?

**Comment by bookshelfdave** [comment]
+1

all tests pass with `make check`

File: db/db_impl.cc (hunk 1 above)
Comment (matthewvon) [inline db/db_impl.cc:1567]: unrelated bug fix ... saw it was wrong, fixed it.

File: port/port_posix.h (hunk 2 above)
Comment (matthewvon) [inline port/port_posix.h:114]: I copied the Mutex class, then changed the names to Spin.  I need spin to be 100% compatible with Mutex ... since 1) I am replacing Mutex with Spin in several places and 2) Spin becomes Mutex on platforms that do not have pthread_spinlock_t.
